File: 3743.maximize-cyclic-partition-score.cpp
#
# @lc app=leetcode id=3743 lang=cpp
#
# [3743] Maximize Cyclic Partition Score
#

# @lc code=start
class Solution {
public:
    long long maximumScore(vector<int>& nums, int k) {
        int n = nums.size();
        long long max_score = 0;
        // Dynamic programming table for storing maximum scores based on partition points
        vector<vector<long long>> dp(n, vector<long long>(k + 1, 0));
        // Iterate over each possible starting point in original `nums`
        for (int start = 0; start < n; ++start) {
            int max_val = INT_MIN, min_val = INT_MAX;
            // Use deque for efficient range queries (maintaining max/min)
            deque<int> maxDeque, minDeque;
            // Extend window up to n elements from start (handle wrap-around using modulo)
            for (int end = start; end < start + n; ++end) {
                int idx = end % n;
                // Maintain max/min using deque
                while (!maxDeque.empty() && nums[maxDeque.back()] <= nums[idx]) {
                    maxDeque.pop_back();
                }
                while (!minDeque.empty() && nums[minDeque.back()] >= nums[idx]) {
                    minDeque.pop_back();
                }
                maxDeque.push_back(idx);
                minDeque.push_back(idx);
                // Calculate range from deques' front elements
                int currentRange = nums[maxDeque.front()] - nums[minDeque.front()];
                if ((end - start + 1) <= k) { // Valid partition size check
                    dp[start][end - start + 1] = std::max(dp[start][end - start + 1], currentRange);
                    if (start > 0) {
                        dp[start][end - start + 1] += dp[start - 1][k];
                    }
                    max_score = std::max(max_score, dp[start][end - start + 1]);
                }
                // Remove out-of-window elements from deques
                if (!maxDeque.empty() && maxDeque.front() == idx - k) {
                    maxDeque.pop_front();
                }
                if (!minDeque.empty() && minDeque.front() == idx - k) {
                    minDeque.pop_front();
                }
            }
        }
        return max_score;
    }
};
# @lc code=end

File: 3743.maximize-cyclic-partition-score.java
#
# @lc app=leetcode id=3743 lang=java
#
# [3743] Maximize Cyclic Partition Score
#

# @lc code=start
class Solution {
    public long maximumScore(int[] nums, int k) {
        int n = nums.length;
        long maxScore = 0;
        int[] arr = new int[2 * n];
        for (int i = 0; i < 2 * n; ++i) arr[i] = nums[i % n];
        // Precompute min and max for all subarrays of length <= n
        long[][] minTable = new long[2 * n][n + 1];
        long[][] maxTable = new long[2 * n][n + 1];
        for (int i = 0; i < 2 * n; ++i) {
            minTable[i][0] = Integer.MAX_VALUE;
            maxTable[i][0] = Integer.MIN_VALUE;
            for (int len = 1; len <= n && i + len <= 2 * n; ++len) {
                if (len == 1) {
                    minTable[i][len] = arr[i];
                    maxTable[i][len] = arr[i];
                } else {
                    minTable[i][len] = Math.min(minTable[i][len - 1], arr[i + len - 1]);
                    maxTable[i][len] = Math.max(maxTable[i][len - 1], arr[i + len - 1]);
                }
            }
        }
        // Try all starting points
        for (int start = 0; start < n; ++start) {
            // DP: dp[i][j] = max score for prefix length i (from start), with j partitions
            long[][] dp = new long[n + 1][k + 1];
            for (int i = 0; i <= n; ++i) for (int j = 0; j <= k; ++j) dp[i][j] = Long.MIN_VALUE;
            dp[0][0] = 0;
            for (int i = 1; i <= n; ++i) {
                for (int part = 1; part <= Math.min(i, k); ++part) {
                    for (int last = part - 1; last < i; ++last) {
                        long range = maxTable[start + last][i - last] - minTable[start + last][i - last];
                        if (dp[last][part - 1] != Long.MIN_VALUE) {
                            dp[i][part] = Math.max(dp[i][part], dp[last][part - 1] + range);
                        }
                    }
                }
            }
            // Get the max for this rotation
            for (int part = 1; part <= k; ++part) {
                maxScore = Math.max(maxScore, dp[n][part]);
            }
        }
        return maxScore;
    }
}
# @lc code=end